### PR TITLE
Feature/sg 217 refactor dataset interface

### PR DESCRIPTION
**Pull request by shaydeci**
- Support for data loaders passed through train
- Dataloader factory + example for coco2017 with defaults defined by yaml file.

### CODE REVIEW

File: docs/_modules/super_gradients/training/sg_model/sg_model.html [hunk 1]
```diff
@@ -104,7 +104,7 @@ <h1>Source code for super_gradients.training.sg_model.sg_model</h1><div class="h
 <span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
 <span class="kn">from</span> <span class="nn">piptools.scripts.sync</span> <span class="kn">import</span> <span class="n">_get_installed_distributions</span>
 
-<span class="kn">from</span> <span class="nn">super_gradients.common.factories.callbacks_factory</span> <span class="kn">import</span> <span class="n">CallbacksFactory</span>
```
Comment (ofrimasad) [inline docs/_modules/super_gradients/training/sg_model/sg_model.html:107]: are we supposed to change those? aren't they generated automatically?

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        cfg.dataset_params.train_dataloader_params or cfg.dataset_params.valid_dataloader_params as defaults for the dataset constructor
+     and
+        cfg.dataset_params.train_dataset_params or cfg.dataset_params.valid_dataset_params as defaults for DataLoader contructor.
+
+    :param dataset_params: dataset params that override the yaml configured defaults, then passed to the dataset_cls.__init__.
+    :param dataloader_params: DataLoader params that override the yaml configured defaults, then passed to the DataLoader.__init__
+    :return: DataLoader
+    """
+    GlobalHydra.instance().clear()
+    with initialize(config_path="../../recipes"):
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:32]: use pkg_resources.resource_filename

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        # config is relative to a module
+        cfg = compose(config_name=config_name)
+
+        _process_dataset_params(cfg, dataset_params, train)
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:36]: if you are changing dataset_params, do not hide it. 
it is unclear from this function call that you change it.
make the function return a dictionary 
```dataset_params = _process_dataset_params(cfg, dataset_params, train)```

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        local_rank = get_local_rank()
+        with wait_for_the_master(local_rank):
+            dataset = dataset_cls(**dataset_params)
+
+        _process_dataloader_params(cfg, dataloader_params, dataset, train)
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:42]: same here:
```dataloader_params = _process_dataloader_params(cfg, dataloader_params, dataset, train)```

**Review by ofrimasad — COMMENTED** [review]
generally, if a function is changing a dictionary, it is better to make that function return a value (even if that is the same dictionary you pass in the input).
that way, when someone reads the code it is clear to them that the value changes 
``` my_dict = foo(a, b, my_dict)```

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 2 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py:34]: not possible here, Hydra expects relative path.

File: docs/_modules/super_gradients/training/sg_model/sg_model.html (hunk 1 above)
Comment (shaydeci) [reply docs/_modules/super_gradients/training/sg_model/sg_model.html:107]: removed

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 3 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py]: https://github.com/Deci-AI/super-gradients/pull/328/commits/14a033e843e97b516050f1326c0106ba3cbd64c5 added.

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 4 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py]: https://github.com/Deci-AI/super-gradients/pull/328/commits/14a033e843e97b516050f1326c0106ba3cbd64c5 added.

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 2 above)
Comment (ofrimasad) [reply src/super_gradients/training/dataloaders/dataloader_factory.py:34]: will it work when installing from pip?

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        batch_size = dataloader_params.pop("batch_size")
+        dataloader_params["batch_sampler"] = BatchSampler(sampler=sampler, batch_size=batch_size, drop_last=False)
+
+    return dataloader_params
+
+
+def _override_default_params_without_nones(params, default_params):
+    for key, val in default_params.items():
+        if key not in params.keys() or params[key] is None:
+            params[key] = val
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:86]: return value here as well

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def _instantiate_sampler(dataset, dataloader_params):
+    sampler_name = list(dataloader_params["sampler"].keys())[0]
+    dataloader_params["sampler"][sampler_name]["dataset"] = dataset
+    dataloader_params["sampler"] = SamplersFactory().get(dataloader_params["sampler"])
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:92]: return value here as well

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    return dataset_params
+
+
+def _process_dataloader_params(cfg, dataloader_params, dataset, train):
+    default_dataloader_params = cfg.dataset_params.train_dataloader_params if train else cfg.dataset_params.val_dataloader_params
+    default_dataloader_params = hydra.utils.instantiate(default_dataloader_params)
+    is_dist = super_gradients.is_distributed()
+
+    if get_param(default_dataloader_params, "sampler") is not None:
+        _instantiate_sampler(dataset, default_dataloader_params)
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:66]: default_dataloader_params = ...
lines 69, 72 as well

File: src/super_gradients/training/dataloaders/dataloader_factory.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    elif is_dist:
+        default_dataloader_params["sampler"] = {"DistributedSampler": {}}
+        _instantiate_sampler(dataset, default_dataloader_params)
+
+    if get_param(dataloader_params, "sampler") is not None:
+        _instantiate_sampler(dataset, dataloader_params)
+
+    _override_default_params_without_nones(dataloader_params, default_dataloader_params)
```
Comment (ofrimasad) [inline src/super_gradients/training/dataloaders/dataloader_factory.py:74]: default_dataloader_params = ...

**Review by ofrimasad — COMMENTED** [review]
some last minor comments

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 2 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py]: Added hydra initialisation with absolute path in https://github.com/Deci-AI/super-gradients/pull/328/commits/0017eedb1ad94dbb6e06bb3c254c6c1eaf9a9ea3

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py:87]: done in https://github.com/Deci-AI/super-gradients/commit/0017eedb1ad94dbb6e06bb3c254c6c1eaf9a9ea3

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 6 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py:94]: done in https://github.com/Deci-AI/super-gradients/commit/0017eedb1ad94dbb6e06bb3c254c6c1eaf9a9ea3 .

File: src/super_gradients/training/dataloaders/dataloader_factory.py (hunk 7 above)
Comment (shaydeci) [reply src/super_gradients/training/dataloaders/dataloader_factory.py]: https://github.com/Deci-AI/super-gradients/commit/0017eedb1ad94dbb6e06bb3c254c6c1eaf9a9ea3 done.

**Review by ofrimasad — APPROVED** [review]
LGTM
